### PR TITLE
scipy in Lambdify  #7229

### PR DESCRIPTION
**Pull request by Shekharrajak**
[7229](https://github.com/sympy/sympy/issues/7229)
- Addition of namespaces ,namespaces_default,translation for scipy in lambdify.py
- added tests in tests/test_lambdify.py (some of them are from previous old PRs)
- updated importtools.py for scipy
- and some basic modifications.

### CODE REVIEW

File: sympy/utilities/tests/test_lambdify.py [hunk 1]
```diff
@@ -9,6 +9,14 @@
 from sympy.utilities.pytest import skip
 from sympy.utilities.decorator import conserve_mpmath_dps
 from sympy.external import import_module
+# for new test cases of scipy
+from sympy.functions.special.bessel import *
+from sympy.functions.special.hyper import * 
+from sympy.functions.special.hyper import *
+from sympy.functions.special.gamma_functions import *
+from sympy.functions.special.gamma_functions import *
```
Comment (moorepants) [inline sympy/utilities/tests/test_lambdify.py]: Please don't use the `from x import *` style of import. Always make it explicit.

File: sympy/utilities/tests/test_lambdify.py [hunk 2]
```diff
@@ -150,17 +159,46 @@ def test_numpy_transl():
...
+    assert f(1) == 1
+
+
+# def test_numpy_translation_():
+#     if not numpy:
+#         skip("numpy is not installed.")
+
+#     f = lambdify(x, Abs(x), "numpy")
+#     assert f(-1) == 1
+#     assert f(1) == 1
```
Comment (moorepants) [inline sympy/utilities/tests/test_lambdify.py]: What is this commented code?

Comment (Shekharrajak) [inline sympy/utilities/tests/test_lambdify.py]: I found same code with different def name in [line : 172](https://github.com/sympy/sympy/pull/10319/files#diff-2220ba8f6c6427dab59fd14c133d009aR172) If it pass/fail first one ,then I don't think we need second def.

File: sympy/utilities/tests/test_lambdify.py (hunk 1 above)
Comment (Shekharrajak) [inline sympy/utilities/tests/test_lambdify.py]: Done

**Comment by Shekharrajak** [comment]
Why this error : 
```
func0 = lambdify((), expr, modules="mpmath", printer=intervalrepr)
NameError: global name 'intervalrepr' is not defined
```
[line](https://github.com/sympy/sympy/blob/master/sympy/utilities/tests/test_lambdify.py#L561)

**Comment by moorepants** [comment]
Do you have it defined? Please post your whole code snippet that you are running? Or are you running the test?

**Comment by Shekharrajak** [comment]
@moorepants , During running the tests in Travis that error is occurring.I did not define,that line is already there.

File: sympy/external/tests/test_scipy.py [hunk 3]
```diff
@@ -24,12 +24,12 @@ def eq(a, b, tol=1e-6):
 
 def test_jn_zeros():
     assert eq(jn_zeros(0, 4, method="scipy"),
-            [3.141592, 6.283185, 9.424777, 12.566370])
+            [3.141592, 6.283185, 9.424777, 12.566370]) == True
```
Comment (moorepants) [inline sympy/external/tests/test_scipy.py]: Why are these changed?

File: sympy/utilities/lambdify.py [hunk 4]
```diff
@@ -88,22 +90,74 @@
     "Matrix": "array",
     "MutableDenseMatrix": "array",
     "ImmutableMatrix": "array",
+    "ImmutableDenseMatrix": "array"
+
+}
+
+# most of the functions of scipy is similar to numpy. Checked each functions.
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Can you simply copy the NumPy dictionary and then update with the additional SciPy functions to reduce the repeated code?

File: sympy/utilities/lambdify.py [hunk 5]
```diff
@@ -88,22 +90,74 @@
...
+    "ceiling": "ceil",
+    "E": "e",
+    "im": "imag",
+    "ln": "log",
+    "Max": "amax",
+    "Min": "amin",
+    "Mod": "mod",
+    "oo": "inf",
+    "re": "real",
+    "SparseMatrix": "array",
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Is this the right thing to do? SciPy has a sparse matrix module that should be used.

File: sympy/utilities/lambdify.py [hunk 6]
```diff
@@ -88,22 +90,74 @@
...
+    "bessely": "yn",
+    "besseli": "iv",
+    "besselk": "kn",
+    "loggamma": "gammaln",
+    "digamma": "psi",
+    "hyper": "hyp1f1",
 }
 
-NUMEXPR_TRANSLATIONS = {}
+NUMEXPR_TRANSLATIONS = {
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: What does this have to do with the SciPy changes? Please make a different PR for this. You'll need some tests for it too.

File: sympy/utilities/lambdify.py [hunk 7]
```diff
@@ -88,22 +90,74 @@
...
+    "im": "imag",
+}
 
 # Available modules:
 MODULES = {
     "math": (MATH, MATH_DEFAULT, MATH_TRANSLATIONS, ("from math import *",)),
     "mpmath": (MPMATH, MPMATH_DEFAULT, MPMATH_TRANSLATIONS, ("from mpmath import *",)),
     "numpy": (NUMPY, NUMPY_DEFAULT, NUMPY_TRANSLATIONS, ("import_module('numpy')",)),
+    # according to the external/importtools.py we should use import_module() for numpy,scipy
+    "scipy": (SCIPY, SCIPY_DEFAULT, SCIPY_TRANSLATIONS, ("import_module('scipy')",)),
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Are all the functions in the top level module import for SciPy?

File: sympy/utilities/lambdify.py [hunk 8]
```diff
@@ -88,22 +90,74 @@
...
     "sympy": (SYMPY, SYMPY_DEFAULT, {}, (
         "from sympy.functions import *",
         "from sympy.matrices import *",
         "from sympy import Integral, pi, oo, nan, zoo, E, I",)),
-    "numexpr" : (NUMEXPR, NUMEXPR_DEFAULT, NUMEXPR_TRANSLATIONS,
-                 ("import_module('numexpr')", )),
+    "numexpr": (NUMEXPR, NUMEXPR_DEFAULT, NUMEXPR_TRANSLATIONS, ("import_module('numexpr')", ))
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Is this under 79 chars?

File: sympy/utilities/lambdify.py [hunk 9]
```diff
@@ -112,13 +166,12 @@ def _import(module, reload="False"):
     Creates a global translation dictionary for module.
 
     The argument module has to be one of the following strings: "math",
-    "mpmath", "numpy", "sympy".
+    "mpmath", "numpy", "sympy", "scipy".
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: What order of precedence do you have here? I think it should go along side NumPy.

File: sympy/utilities/lambdify.py [hunk 10]
```diff
@@ -112,13 +166,12 @@ def _import(module, reload="False"):
...
     These dictionaries map names of python functions to their equivalent in
     other modules.
     """
     try:
-        namespace, namespace_default, translations, import_commands = MODULES[
-            module]
+        namespace, namespace_default, translations, import_commands = MODULES[module]
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Please keep things under 79 chars.

File: sympy/utilities/lambdify.py [hunk 11]
```diff
@@ -155,8 +208,8 @@ def _import(module, reload="False"):
 
 
 @doctest_depends_on(modules=('numpy'))
-def lambdify(args, expr, modules=None, printer=None, use_imps=True,
-        dummify=True):
+@doctest_depends_on(modules=('scipy'))
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Why did you remove the dependence on NumPy? Because SciPy depends on NumPy?

File: sympy/utilities/lambdify.py [hunk 12]
```diff
@@ -155,8 +208,8 @@ def _import(module, reload="False"):
...
+def lambdify(args, expr, modules=None, printer=None, use_imps=True, dummify=True):
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Keep to 79 chars.

File: sympy/utilities/lambdify.py [hunk 13]
```diff
@@ -165,7 +218,7 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
     functions - exactly in this order. To change this behavior, the "modules"
     argument can be used. It accepts:
 
-     - the strings "math", "mpmath", "numpy", "numexpr", "sympy"
+     - the strings "math", "mpmath", "numpy", "numexpr", "sympy" ,"scipy"
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Please justify this order of modules.

File: sympy/utilities/lambdify.py [hunk 14]
```diff
@@ -233,6 +286,19 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
...
         [-2.18503986 -0.29100619 -0.8559934 ]
 
+        Use of scipy module:
+
+        >>> import scipy
+        >>> import numpy
+        >>> a = numpy.arange(10)
+        >>> a
+        [0 1 2 3 4 5 6 7 8 9]
+        >>> expr=sin(x)
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Please check your code with a linter that checks for PEP8 issues, e.g. spaces around the `=` here.

File: sympy/utilities/lambdify.py [hunk 15]
```diff
@@ -233,6 +286,19 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
...
+        >>> f= lambdify(x,expr,scipy)
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: I think it is preferable to use the sting version of modules. This version should really only be used for custom modules.

File: sympy/utilities/lambdify.py (hunk 14 above)
Comment (moorepants) [inline sympy/utilities/lambdify.py]: It'd be preferable to use a function in this example that is unique to SciPy.

File: sympy/utilities/lambdify.py [hunk 16]
```diff
@@ -335,18 +416,22 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
         namespace.update(buf)
 
     if hasattr(expr, "atoms"):
-        #Try if you can extract symbols from the expression.
-        #Move on if expr.atoms in not implemented.
+        #   Try if you can extract symbols from the expression.
+        # Move on if expr.atoms in not implemented.
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Please keep all of these kinds of fixes in a separate PR (or at least separate commit). I makes it difficult to focus on the actual code changes.

File: sympy/utilities/lambdify.py [hunk 17]
```diff
@@ -360,8 +445,7 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
             names.append(var.name)
         else:
             # It's an iterable. Try to get name by inspection of calling frame.
-            name_list = [var_name for var_name, var_val in callers_local_vars
-                    if var_val is var]
+            name_list = [var_name for var_name, var_val in callers_local_vars if var_val is var]
```
Comment (moorepants) [inline sympy/utilities/lambdify.py]: 79 chars.

File: sympy/utilities/tests/test_lambdify.py [hunk 18]
```diff
@@ -4,11 +4,20 @@
...
 from sympy.printing.lambdarepr import LambdaPrinter
+from sympy.printing.printer import Printer
 import mpmath
 from sympy.utilities.lambdify import implemented_function
 from sympy.utilities.pytest import skip
 from sympy.utilities.decorator import conserve_mpmath_dps
 from sympy.external import import_module
+# for new test cases of scipy
+from sympy.functions.special.bessel import besselj, bessely, besseli, besselk, hankel1, hankel2
+# from sympy.functions.special.hyper import hyper
```
Comment (moorepants) [inline sympy/utilities/tests/test_lambdify.py]: Why is this commented out?

File: sympy/utilities/tests/test_lambdify.py [hunk 19]
```diff
@@ -150,17 +160,46 @@ def test_numpy_transl():
         assert nump in numpy.__dict__
 
 
+def test_scipy_transl():
+    if not scipy:
+        skip("scipy is not installed.")
+
+    from sympy.utilities.lambdify import SCIPY_TRANSLATIONS
```
Comment (moorepants) [inline sympy/utilities/tests/test_lambdify.py]: Can this go in the top of the file?

File: sympy/utilities/tests/test_lambdify.py (hunk 2 above)
Comment (moorepants) [inline sympy/utilities/tests/test_lambdify.py]: It is it not needed remove it. Don't just commented it out.

File: sympy/utilities/tests/test_lambdify.py [hunk 20]
```diff
@@ -581,3 +629,71 @@ def test_issue_2790():
...
     assert lambdify((x, y, z), ITE(x, y, z))(False, 5, 3) == 3
+
+
+def test_scipy_translation_bessel():
+    if not scipy:
+        skip("scipy not installed.")
+
+    f = lambdify((x, y), besselj(x,y), "scipy")
+    assert f(1, 2) == 0.5767248077568734
+    assert f(3, 4) == 0.43017147387562199
```
Comment (moorepants) [inline sympy/utilities/tests/test_lambdify.py]: This is not a good way to test floating point numbers. Please use NumPy's assert_allclose or similar functions.

**Comment by moorepants** [comment]
All of the formatting changes make it difficult to see the code changes. It is much more preferable and easier for the reviewers if you do formatting in a different PR.

File: sympy/external/tests/test_scipy.py (hunk 3 above)
Comment (Shekharrajak) [inline sympy/external/tests/test_scipy.py]: test cases returns `True` so I think it will be better if we write it clearly., what is the output.

File: sympy/utilities/lambdify.py (hunk 4 above)
Comment (Shekharrajak) [inline sympy/utilities/lambdify.py]: I checked all the functions with both scipy and numpy.These are working for both , so these should be similar and I added some additioal functions for Scipy.

File: sympy/utilities/lambdify.py (hunk 5 above)
Comment (Shekharrajak) [inline sympy/utilities/lambdify.py]: I am not sure about this(I searched in internet, but didn't find) , but because array is used in Scipy and numpy very similarly,so I thought we can map them.

File: sympy/utilities/lambdify.py (hunk 6 above)
Comment (Shekharrajak) [inline sympy/utilities/lambdify.py]: Oh! actually I wastrying some examples which used both scipy,numexpr and numpy,numexpr  , but I was getting some errors that time I thought mapping should be there for numexpr also.I will remove them and make different PR for this.

File: sympy/utilities/lambdify.py (hunk 10 above)
Comment (Shekharrajak) [inline sympy/utilities/lambdify.py]: Okay.

File: sympy/utilities/tests/test_lambdify.py (hunk 18 above)
Comment (Shekharrajak) [inline sympy/utilities/tests/test_lambdify.py]: I think I should remove this.there is no test case that use hyper.

File: sympy/external/tests/test_scipy.py (hunk 3 above)
Comment (moorepants) [inline sympy/external/tests/test_scipy.py]: The convention (see PEP8) is to not compare to True like this. It is unnecessary. Please revert.

File: sympy/utilities/lambdify.py (hunk 4 above)
Comment (moorepants) [inline sympy/utilities/lambdify.py]: I don't think that there is any reason to have these doubled up. Can you do something like:

```python
NUMPY_TRANSLATIONS = {...}
SCIPY_TRANSLATIONS = {...}
SCIPY_TRANSLATIONS.update(NUMPY_TRANSLATIONS)
```

The SciPy namespace imports the numpy namspace so there shouldn't be a reason to write these duplicates.

File: sympy/utilities/lambdify.py (hunk 5 above)
Comment (moorepants) [inline sympy/utilities/lambdify.py]: I don't think we should convert sympy SparseMatrices to simple NumPy arrays. If you convert them to anything it should be to http://docs.scipy.org/doc/scipy/reference/sparse.html.

File: sympy/utilities/lambdify.py (hunk 6 above)
Comment (moorepants) [inline sympy/utilities/lambdify.py]: Ok.

File: sympy/utilities/tests/test_lambdify.py (hunk 18 above)
Comment (moorepants) [inline sympy/utilities/tests/test_lambdify.py]: Please. Thanks.

File: sympy/utilities/lambdify.py (hunk 7 above)
Comment (Shekharrajak) [inline sympy/utilities/lambdify.py]: Import_module import and return a module if it is installed.So I think it imports all the functions just like numpy ,otherwise we need to add `from scipy import *`

File: sympy/utilities/tests/test_lambdify.py (hunk 20 above)
Comment (Shekharrajak) [inline sympy/utilities/tests/test_lambdify.py]: @moorepants Do you mean like this : 
```
assert (numpy.allclose(f(1, 2), [0.5767248077568734, 0.43017147387562199], rtol=1e-05, atol=1e-08)) is True
```?

**Comment by moorepants** [comment]
Use this module for testing: http://docs.scipy.org/doc/numpy-1.10.0/reference/routines.testing.html

**Comment by Shekharrajak** [comment]
Check this [modification ](https://github.com/sympy/sympy/commit/b71f16f476141adb75c9bc5f5bc34b4a188a13da#diff-2220ba8f6c6427dab59fd14c133d009aR629)

File: sympy/utilities/lambdify.py (hunk 7 above)
Comment (asmeurer) [inline sympy/utilities/lambdify.py]: Quite a few are in scipy.special.
